inst/include/interface/rcpp/rcpp_objects/rcpp_tmb_distribution.hpp

```diff
@@ -263,116 +215,79 @@ class DlnormDistributionsInterface : public DistributionsInterfaceBase {
-    fims::Dmultinom<double> dmultinom;
-    // Decale TMBVector in this scope
-    typedef
-        typename fims::ModelTraits<TMB_FIMS_REAL_TYPE>::EigenVector TMBVector;
-    dmultinom.x = TMBVector(x.size());  // Vector from TMB
-    dmultinom.p = TMBVector(p.size());  // Vector from TMB
-    for (int i = 0; i < x.size(); i++) {
-      dmultinom.x[i] = x[i];
-      dmultinom.p[i] = p[i];
+    DmultinomDistributionsInterface() : DistributionsInterfaceBase() {
     }
-    return dmultinom.evaluate(do_log);
-  }
-
-  virtual bool add_to_fims_tmb() {
-    typedef typename fims::ModelTraits<TMB_FIMS_REAL_TYPE>::EigenVector Vector0;
-    std::shared_ptr<fims::Information<TMB_FIMS_REAL_TYPE>> d0 =
-        fims::Information<TMB_FIMS_REAL_TYPE>::GetInstance();
 
-    std::shared_ptr<fims::Dmultinom<TMB_FIMS_REAL_TYPE>> model0 =
-        std::make_shared<fims::Dmultinom<TMB_FIMS_REAL_TYPE>>();
-
-    model0->id = this->id;
-    model0->x = Vector0(x.size());
-    model0->p = Vector0(p.size());
-
-    for (int i = 0; i < x.size(); i++) {
-      model0->x[i] = x[i];
-      model0->p[i] = p[i];
+    virtual ~DmultinomDistributionsInterface() {
     }
 
-    d0->distribution_models[model0->id] = model0;
-
-    typedef
-        typename fims::ModelTraits<TMB_FIMS_FIRST_ORDER>::EigenVector Vector1;
-    std::shared_ptr<fims::Information<TMB_FIMS_FIRST_ORDER>> d1 =
-        fims::Information<TMB_FIMS_FIRST_ORDER>::GetInstance();
-
-    std::shared_ptr<fims::Dmultinom<TMB_FIMS_FIRST_ORDER>> model1 =
-        std::make_shared<fims::Dmultinom<TMB_FIMS_FIRST_ORDER>>();
-
-    model1->id = this->id;
-    model1->x = Vector1(x.size());
-    model1->p = Vector1(p.size());
+    virtual uint32_t get_id() {
+        return this->id;
+    }
 
-    for (int i = 0; i < x.size(); i++) {
-      model1->x[i] = x[i];
-      model1->p[i] = p[i];
+    /**
+     * @brief Evaluate multinom probability density function, default returns the
+     * log of the pdf
+     *
+     * @tparam T
+     * @return log pdf
+     */
+    virtual double evaluate(bool do_log) {
+        fims::Dmultinom<double> dmultinom;
+        // Decale TMBVector in this scope
+        typedef
+        typename fims::ModelTraits<TMB_FIMS_REAL_TYPE>::EigenVector TMBVector;
+        dmultinom.x = TMBVector(x.size()); // Vector from TMB
+        dmultinom.p = TMBVector(p.size()); // Vector from TMB
+        for (int i = 0; i < x.size(); i++) {
+            dmultinom.x[i] = x[i];
+            dmultinom.p[i] = p[i];
+        }
+        return dmultinom.evaluate(do_log);
     }
 
-    d1->distribution_models[model1->id] = model1;
+#ifdef TMB_MODEL
 
-    typedef
-        typename fims::ModelTraits<TMB_FIMS_SECOND_ORDER>::EigenVector Vector2;
-    std::shared_ptr<fims::Information<TMB_FIMS_SECOND_ORDER>> d2 =
-        fims::Information<TMB_FIMS_SECOND_ORDER>::GetInstance();
+    template<typename T>
+    bool add_to_fims_tmb_internal() {
+        typedef typename fims::ModelTraits<T>::EigenVector Vector0;
```

Comment:
question: why do we define `typedef typename Eigenvector` as type `TMBVector` above and `Vector0` in this scope?